Use D-Bus for Linux theme awareness

J3RN opened the pull request:
The thrust of this change is to use D-Bus' notion of "color-scheme" instead of Dconf's notion of GTK theme, as detailed in #19.

Feature changes:
- Naturally `auto-dark--is-dark-mode-dconf` has become `auto-dark--is-dark-mode-dbus`.  The new one (D-Bus) uses Emacs' built-in D-Bus integration.
- Add register and unregister functions for a D-Bus listener.  (`auto-dark--register-dbus-listener` and `auto-dark--unregister-dbus-listener`).  This necessitated the creation of the `auto-dark--dbus-listener-object` variable as a side-effect of how Emacs' D-Bus integration functions.

Other changes:
- Reduced duplication by changing `auto-dark--check-and-set-dark-mode` to call `auto-dark--set-theme` instead of duplicating the setting logic.
- Generalized the [un]registering of listeners into `auto-dark--[un]register-change-listener` functions that semi-intelligently dispatch to either `ns-*` functions for macOS or `dbus-*` functions for Linux.

I have some more ideas for cleaning up the macOS stuff, but I'll save those for another PR :grin:

LionyxML reviewed (approved):
Thanks for your great additions!

As a side-note, byte compiling this under macOS now gives 3 warnings:
```
auto-dark.el:168:4: Warning: the function ‘dbus-unregister-object’ is not
    known to be defined.
auto-dark.el:154:10: Warning: the function ‘dbus-register-signal’ is not known
    to be defined.
auto-dark.el:80:16: Warning: the function ‘dbus-call-method’ is not known to
    be defined.
```

I am guessing this is fine since most probably macOS builds should not have "d-bus" support and there's no flow leading to it, unless the OS is GNU/Linux.

LionyxML commented:
@seagle0128 , since you were the first to implement something for Linux, I would appreciate some feedback on this take by @J3RN.

There was some discussion on Reddit and here on #19  about d-bus being a better way to check the theme state then relying on the current method that some said only worked with a specific theme.

seagle0128 commented on auto-dark.el:
```diff
@@ -148,19 +146,67 @@ Argument APPEARANCE should be light or dark."
-        (remove-hook 'ns-system-appearance-change-functions #'auto-dark--ns-set-theme)
-      (auto-dark-stop-timer))))
+      (progn
+	(auto-dark--check-and-set-dark-mode)
```
Please format the codes.

J3RN replied:
As far as I'm aware, there's no standard formatting for Emacs Lisp.  There are a few opinionated packages, e.g. [elisp-format](https://github.com/Yuki-Inoue/elisp-format) and [elfmt](https://github.com/riscy/elfmt), but they vary widely on what they consider to be "formatted."

Is it the tabs that you don't like?  I originally kept the tabs as Emacs inserted them thinking that this was the convention for this file, but I see now that originally they were only used in the Applescript in `auto-dark--is-dark-mode-builtin`.  I'd be happy to convert them to spaces if that's your preference.

LionyxML replied:
Personally I haven't found yet a config for elisp-format that pleases me :)

Originally I've used both spaces and tabs as `indent-tabs-mode` was "on" for elisp-mode and  I did not noticed that before.

That said, I tend to a `spaces only` approach in other languages. What do you think we should do here?

seagle0128 commented:
Basically dbus is a more common approach, but I didn't test with many Linux distributions. I agree with the changes, but please format the codes before check-in.

LionyxML commented on auto-dark.el:
```diff
@@ -148,19 +146,67 @@ Argument APPEARANCE should be light or dark."
     (cancel-timer auto-dark--timer)))
 
+(defun auto-dark--register-dbus-listener ()
+  "Ask D-Bus to send us a signal on theme change and add a callback
```
Hello again @J3RN! Reopening this review since I should've oriented you better (my bad) :)

Part of keeping a package on MELPA is following this rules:
- Should clean pass byte compiling
- Should be OK with `checkdock` 

<img width="450" alt="image" src="https://user-images.githubusercontent.com/16169950/210099311-3d6acb57-40c1-4dc4-ab46-51abc9d50588.png">

At this line, checkdoc complains the first line is not a complete sentence. It means that the first line should be a short sentence ending with period (.). Two or more lines are allowed, but the first one should be a complete short sentence.

Since both functions have the same comment, something to differ them would be better, like:
```
(defun auto-dark--register-dbus-listener ()
  "Register a listener function to D-Bus.
Ask D-Bus to send us a signal on theme change and add a callback
to change the theme."
```

and 

```
(defun auto-dark--unregister-dbus-listener ()
  "Unregister a listener function to D-Bus.
Ask D-Bus to send us a signal on theme change and add a callback
to change the theme."
```

LionyxML commented on auto-dark.el:
```diff
@@ -148,19 +146,67 @@ Argument APPEARANCE should be light or dark."
+             (auto-dark--set-theme (pcase (car val) (0 'light) (1 'dark))))))))
+
+(defun auto-dark--unregister-dbus-listener ()
+  "Ask D-Bus to send us a signal on theme change and add a callback
```
See comment on line 150.

LionyxML reviewed (changes requested):
Thanks again! This will be of great value for the package. 

I'm reopening the review so we can polish it a little bit further.

You can also add your name to the top of the .el file as we're solving this minor issues.

J3RN commented:
I rewrote the doc strings such that `checkdoc` no longer complains and I converted all indentation to spaces except where it was tabs before (i.e. the Applescript) as the tabs seemed rather arbitrarily placed.

This byte-compiles cleanly on my machine, except for a gripe about `ns-do-applescript`, which is present in [Emacs for Mac OS X](https://emacsformacosx.com/) and [Emacs Plus](https://github.com/d12frosted/homebrew-emacs-plus).  IMO probably nothing to worry about.

LionyxML commented:
Thanks you all!

It looks like we did it. A 31th Dec merge :)

real-or-random reviewed (commented):
Oh awesome! I hadn't seen this PR earlier, but here's a post-merge review

The most important comment:

I need a `(require 'dbus)` before enabling auto-dark-mode.

seagle0128 commented on auto-dark.el:
```diff
@@ -72,9 +75,14 @@ end tell")))
-  (string-equal "'Yaru-dark'\n" (string-trim (shell-command-to-string "dconf read /org/gnome/desktop/interface/gtk-theme"))))
+(defun auto-dark--is-dark-mode-dbus ()
+  "Use Emacs built-in D-Bus function to determine if dark theme is enabled."
+  (eq 1 (caar (dbus-call-method
```
Maybe the better solution is `(when (require 'dbus nil t)` .

real-or-random replied:
How would this avoid the timeout?

LionyxML replied:
I think @seagle0128 meant to address this quote:
![image](https://user-images.githubusercontent.com/16169950/210254965-f3c10a45-fb28-422f-a20a-92d3ebd64e47.png).

real-or-random replied:
Oh sure!

edgar-vincent commented:
Thank you very much for this PR, @J3RN, it's lovely!
